CI_README.md

```diff
@@ -0,0 +1,36 @@
+
+# CI For CI Developers
+If you're working on enhancing these CI tools, you'll need to know about the structure of the pipeline and about how to develop and test.
+
+## How the pipeline works
+The pipeline is written in Github Actions, and is defined in the workflow .yml files in the .github/workflows directory of this repository.  The documentation on those files is located [here](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/building-actions).  Documentation on the substeps is located below.
+
+### Generation & Diffing
+The generation / diff pipeline has one Action per downstream.  It generates the downstream at the PR's merge commit and at the left-side parent (`HEAD~`), which guarantees isolation of exclusively the changes made by the PR in question.  It creates commits for the downstreams as before-and-after commits.  Those commits don't have relevant git history - they're not meant to be applied - but they are used in a [two-dot diff](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-comparing-branches-in-pull-requests#three-dot-and-two-dot-git-diff-comparisons).  This means that if there are no changes to the downstream, you'll see an empty diff.  These downstream branches are named `pr-$number-old` and `pr-$number-new` respectively.
+
+### Downstream Pushing
+The downstream pushing pipeline runs on a cron job - every 20 minutes.  It checks a branch called `downstream-master` to find the commit which was most recently pushed to downstreams, then collects all commits since then.  One at a time, it generates each downstream at each commit, building an in-order history.  It pushes all the downstreams directly to master branches, but it does not use the `--force` flag - if the branch goes out of sync, the push will fail without damaging the downstream repository history.
+
+## Deploying the pipeline
+The code on the `master` branch is used in all cases - if you are making changes to the Actions, your changes will not apply until they are merged in.
+
+## Design choices & tradeoffs
+* The downstreams share some setup code in common - especially TPG and TPGB.  We violated the DRY principle by writing separate workflows for each repo.  In practice, this has substantially reduced the amount of code - the coordination layer above the two repos was larger than the code saved by combining them.  We also increase speed, since each Action runs separately.
+* The downstream push doesn't happen immediately.  We caused some delay with the cron-based approach that could have been avoided.  This ensures that two commits which are merged at about the same time will never conflict, since only one copy of the downstream push will be running at a time.
```

Comment:
What if the downstream push takes > 20 minutes?

Reply:
That would be sad!  the second one will either
a) finish after the first one and fail to push since it's non-fast-forward, no big deal.
b) finish before the first one, having completed the entire backlog, and the first one will fail since it's non-fast-forward.
c) finish at exactly the same time as the first one, leading to both of them failing, and the next cycle will have to work.

We'll eventually clear any jams, seems to me.